text/3617-precise-capturing.md

```diff
@@ -0,0 +1,670 @@
+
+## Inferred precise capturing
+
+We had hoped that we might be able to achieve something with a similar effect to precise capturing at the cost of an extra generic lifetime parameter in each signature with improvements to the type system.  The goal would be to allow, e.g., this code to work rather than error:
+
+```rust
+fn foo<'o, T>(_: T) -> impl Sized + 'o {}
+
+fn bar(x: ()) -> impl Sized + 'static {
+    foo(&x)
+//~^ ERROR returns a value referencing data owned by the
+//~|       current function
+}
+```
+
+The idea is that, even though the opaque type returned by `foo` does capture the generic type parameter `T`, since the opaque type is explicitly bounded by `'o` and the signature does not assert `T: 'o`, we know that the hidden type cannot actually use `T`.
+
+As it turns out, making full use of this observation is challenging (see [#116040][] and [#116733][]).  While we did make improvements to the type system here, and while more might be possible, this does not solve the problem today in all important cases (including, e.g., avoiding the capture of higher ranked lifetimes in nested opaque types) and will not for the foreseeable future.
+
+Moreover, even with the fullest possible version of these improvements, whether or not a generic parameter is captured by an opaque type would remain observable.  Having an explicit syntax to control what is captured is more direct, more expressive, and leads to a better migration story.
+
+See [Appendix G][] in [RFC 3498][] for more details.
+
+[#116040]: https://github.com/rust-lang/rust/pull/116040
+[#116733]: https://github.com/rust-lang/rust/pull/116733
+[Appendix G]: https://github.com/rust-lang/rfcs/blob/master/text/3498-lifetime-capture-rules-2024.md#appendix-g-future-possibility-inferred-precise-capturing
+
+## Syntax
+
+We considered a number of different possible syntaxes before landing on `impl use<..> Trait`.  We'll discuss each considered.
+
+### `impl use<..> Trait`
+
+This is the syntax chosen in this RFC.
+
+Using a separate keyword makes this syntax more scalable in the sense that we can apply `use<..>` in other places.
+
+Conveniently, the word "use" is quite appropriate here, since we are *using* the generic parameters in the type of the opaque type and allowing the generic parameters to be *used* in the hidden type.  That is, with `use`, we are bringing the generic parameters *into scope* for the hidden type, and `use` is the keyword in Rust for bringing things into scope.
+
+Picking an existing keyword allows for this syntax, including extensions to other positions, to be allowed in older editions.  Because `use` is a full keyword, we're not limited in where it can be placed.
+
+By not putting the generic parameters on `impl<..>`, we reduce the risk of confusion that we are somehow introducing generic parameters here rather than using them.
```

Comment:
(I don't feel strongly about this syntax either way)

This would hardly be the only place where syntax for introducing generic parameters is similar to syntax for using them. `fn foo<T>()` vs `foo::<T>()` for example.

Also, there is a symmetry of a sort between the two kinds of `impl` generics. In a trait implementation, they introduce generic parameters available for use by the type and trait; in RPIT, they would introduce the generics available for use by the hidden type.